File: model/src/main/java/io/sparkled/model/render/RenderedFrame.java
package io.sparkled.model.render;

public class RenderedFrame {

    private final int startFrame;
    private final int frameNumber;
    private final int ledCount;
    private final byte[] data;

    /**
     * Default constructor required for Gson.
     */
    @SuppressWarnings("unused")
    public RenderedFrame() {
        this(0, 0, 0, new byte[]{});
    }

    public RenderedFrame(int startFrame, int frameNumber, int ledCount, byte[] data) {
        this.startFrame = startFrame;
        this.frameNumber = frameNumber;
        this.ledCount = ledCount;
        this.data = data;
    }

    public int getFrameNumber() {
        return frameNumber;
    }

    public byte[] getData() {
        int bytesPerFrame = ledCount * Led.BYTES_PER_LED;
        int offset = (frameNumber - startFrame) * bytesPerFrame;

        byte[] frameData = new byte[bytesPerFrame];
        System.arraycopy(data, offset, frameData, 0, bytesPerFrame);

        return frameData;
    }

    public int getLedCount() {
        return ledCount;
    }

    public Led getLed(int ledIndex) {
        int bytesPerFrame = ledCount * Led.BYTES_PER_LED;
        int offset = (frameNumber - startFrame) * bytesPerFrame;
        return new Led(data, ledIndex, offset);
    }
}

File: model/src/main/kotlin/io/sparkled/model/render/RenderedFrame.kt
package io.sparkled.model.render

@Suppress("ArrayInDataClass")
data class RenderedFrame(
    private val startFrame: Int,
    val frameIndex: Int,
    val ledCount: Int,
    private val data: ByteArray, // TODO use ByteBuffer to fix warning.
    private val dummyFrame: Boolean = false,
) {

    fun getData(): ByteArray {
        if (dummyFrame) {
            // Dummy frames should only ever appear in preview renders for stateful effects.
            throw RuntimeException("Attempted to retrieve data from dummy frame.")
        }

        val bytesPerFrame = ledCount * Led.BYTES_PER_LED
        val offset = (frameIndex - startFrame) * bytesPerFrame

        val frameData = ByteArray(bytesPerFrame)
        System.arraycopy(data, offset, frameData, 0, bytesPerFrame)

        return frameData
    }

    fun getLed(pixelIndex: Int): Led {
        return if (dummyFrame) {
            Led(byteArrayOf(0, 0, 0), 0, 0)
        } else {
            val bytesPerFrame = ledCount * Led.BYTES_PER_LED
            val offset = (frameIndex - startFrame) * bytesPerFrame
            Led(data, pixelIndex, offset)
        }
    }
}
